tests/flake8_typing_imports_test.py

```diff
@@ -118,3 +118,76 @@ def test_missing_overload_define():
         )
     with version_ctx(Version(3, 5, 2)):
         assert not results('from typing import overload')
+
+
+@pytest.mark.parametrize(
+    's', [
+        pytest.param(
+            'from typing import Pattern, Union\n'
+            'def foo(bar: Union[Pattern, str]): pass\n',
+            id='Pattern',
+        ),
+        pytest.param(
+            'import typing\n'
+            'def foo(bar: typing.Union[typing.Pattern, str]): pass\n',
+            id='typing.Pattern',
+        ),
+        pytest.param(
+            'from typing import Match, Union\n'
+            'def foo(bar: Union[Match, str]): pass\n',
+            id='Match',
+        ),
+        pytest.param(
+            'import typing\n'
+            'def foo(bar: typing.Union[typing.Match, str]): pass\n',
+            id='typing.Match',
+        ),
+        pytest.param(
+            'from typing import Match, Pattern, Union\n'
+            'def foo(bar: Union[Match, Pattern, int]): pass\n',
+            id='Match and Pattern',
+        ),
+    ],
+)
+def test_union_pattern_or_match(s):
+    with version_ctx(Version(3, 5, 0)):
+        assert results(s) == {
+            '2:13: TYP003 Union[Match, ...] or Union[Pattern, ...] '
+            'must be quoted <3.5.2',
+        }
+
+    with version_ctx(Version(3, 5, 2)):
+        assert not results(s)
+
+
+@pytest.mark.parametrize(
+    's', [
+        pytest.param(
+            'from bar import Bar\n'
+            'def foo(bar: Union[Bar]): pass\n',
+            id='neither Pattern, nor Match',
+        ),
+        pytest.param(
+            'from typing import Pattern, Union\n'
+            'def foo(bar: Union[Pattern]): pass\n',
+            id='single Pattern',
+        ),
+        pytest.param(
+            'from typing import Pattern, Union\n'
+            'def foo(bar: "Union[Pattern, str]"): pass\n',
+            id='quoted Pattern',
+        ),
+        pytest.param(
+            'from typing import Match, Union\n'
+            'def foo(bar: "Union[Match, str]"): pass\n',
+            id='quoted Match',
+        ),
+        pytest.param(
+            'from typing import Pattern as Pttrn, Union\n'
+            'def foo(bar: Union[Pttrn, str]): pass\n',
+            id='Pattern imported as Pttrn',
```

Comment:
in an ideal world we'd detect this as well, but it's impossible to handle _every_ case of this -- also it's a bit nonsensical

Reply:
Should I remove this case?

Reply:
maybe just a comment `# this is broken too, but so unlikely we elect not to detect it`

Reply:
Or move to the above test and mark it xfail?

Reply:
up to you, idrc